### PR TITLE
Fix flow loops for single switch flows

### PR DESCRIPTION
**Pull request by rozdy**
Related to #3846 

MERGE NOTE: please merge after this one #3873

### CODE REVIEW

**Review by dpoltavets — APPROVED** [review]

**Review by nikitamarchenko — APPROVED** [review]
I didn't find anything terrible

File: src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/flow/ingress/of/IngressFlowLoopInstallMultiTableFlowModFactory.java [hunk 1]
```diff
@@ -15,18 +15,41 @@
...
+public class IngressFlowLoopInstallMultiTableFlowModFactory extends IngressInstallFlowModFactory {
     public IngressFlowLoopInstallMultiTableFlowModFactory(
-            IngressFlowSegmentCommand command, IOFSwitch sw, Set<SwitchFeature> features) {
+            IngressFlowSegmentBase command, IOFSwitch sw, Set<SwitchFeature> features) {
         super(new OfFlowModAddMultiTableMessageBuilderFactory(SwitchManager.FLOW_LOOP_PRIORITY), command, sw, features);
     }
+
+    @Override
+    protected List<OFAction> makeServer42IngressFlowTransformActions(List<Integer> vlanStack) {
+        return null;
```
Comment (niksv) [inline src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/flow/ingress/of/IngressFlowLoopInstallMultiTableFlowModFactory.java:38]: why null? why not empty list?

File: src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/flow/ingress/of/IngressFlowLoopInstallMultiTableFlowModFactory.java [hunk 2]
```diff
@@ -15,18 +15,41 @@
...
+    }
+
+    @Override
+    protected List<OFAction> makeTransformActions(List<Integer> vlanStack) {
+        return null;
```
Comment (niksv) [inline src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/flow/ingress/of/IngressFlowLoopInstallMultiTableFlowModFactory.java:43]: same

File: src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/flow/ingress/of/IngressFlowLoopInstallMultiTableFlowModFactory.java [hunk 3]
```diff
@@ -15,18 +15,41 @@
...
+    }
+
+    @Override
+    protected List<OFInstruction> makeMetadataInstructions() {
+        return null;
```
Comment (niksv) [inline src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/flow/ingress/of/IngressFlowLoopInstallMultiTableFlowModFactory.java:48]: why null? why not empty list?

File: src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/flow/ingress/of/IngressFlowLoopInstallMultiTableFlowModFactory.java [hunk 4]
```diff
@@ -15,18 +15,41 @@
...
+    }
+
+    @Override
+    protected OFAction makeOutputAction() {
+        return null;
```
Comment (niksv) [inline src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/flow/ingress/of/IngressFlowLoopInstallMultiTableFlowModFactory.java:53]: As I see this method used as
`applyActions.add(makeOutputAction());`

Will we add null to the list?

**Review by niksv — CHANGES_REQUESTED** [review]

File: src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/flow/ingress/of/IngressFlowLoopInstallMultiTableFlowModFactory.java (hunk 4 above)
Comment (rozdy) [reply src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/flow/ingress/of/IngressFlowLoopInstallMultiTableFlowModFactory.java:53]: All these methods are used only to generate regular flow rules. This factory is designed to create loop rules and should not be used for regular flow rules. So it's doesn't matter null or empty lists are used here. Maybe throw UnsupportedOperationException is the better way.

Comment (niksv) [reply src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/flow/ingress/of/IngressFlowLoopInstallMultiTableFlowModFactory.java:53]: I do not want to block PR, but I wondering about NPEs. I agree with UnsupportedOperationException idea

Comment (rozdy) [reply src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/flow/ingress/of/IngressFlowLoopInstallMultiTableFlowModFactory.java]: Fixed.

**Review by niksv — APPROVED** [review]

File: src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/flow/ingress/of/IngressFlowLoopFlowModFactory.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    @Override
+    protected OFAction makeOutputAction() {
+        return null;
+    }
+
+    @Override
+    protected List<OFInstruction> makeMetadataInstructions() {
+        return Collections.emptyList();
```
Comment (niksv) [inline src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/flow/ingress/of/IngressFlowLoopFlowModFactory.java:57]: should these 4 methods throw unsupported exceptions too?

Comment (rozdy) [reply src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/flow/ingress/of/IngressFlowLoopFlowModFactory.java]: Fixed.

**Pull request merged**